core/domain/html_validation_service.py

```diff
@@ -820,54 +818,149 @@ def regenerate_image_filename_using_dimensions(filename, height, width):
+                image['filepath-with-value'] == '' or
+                image['filepath-with-value'] == '&quot;&quot;'):
             image.decompose()
             continue
 
         try:
-            filename = json.loads(unescape_html(image['filepath-with-value']))
+            filename = json.loads(
+                unescape_html(image['filepath-with-value'].replace('\\"', '')))
             image['filepath-with-value'] = escape_html(json.dumps(
-                get_filename_with_dimensions(filename, exp_id)))
+                get_filename_with_dimensions(file_system, filename)))
         except Exception as e:
+            entity_type = file_system.impl.assets_path.split('/')[0]
+            entity_id = file_system.impl.assets_path.split('/')[1]
             logging.error(
-                'Exploration %s failed to load image: %s' %
-                (exp_id, image['filepath-with-value'].encode('utf-8')))
+                '%s %s failed to load image: %s' % (
+                    entity_type, entity_id,
+                    image['filepath-with-value'].encode('utf-8')))
             raise e
     return python_utils.UNICODE(soup).replace('<br/>', '<br>')
 
 
-def get_filename_with_dimensions(old_filename, exp_id):
+def add_dimensions_to_image_tags_inside_tabs_and_collapsible_blocks(
+        file_system, html_string):
+    """Adds dimensions to all oppia-noninteractive-image tags inside tabs and
+    collapsible blocks. Removes image tags that have no filepath.
+
+    Args:
+        file_system: GcsFileSystem. File system of the entity.
+        html_string: str. HTML string to modify.
+
+    Returns:
+        str. Updated HTML string with the dimensions for all
+            oppia-noninteractive-image tags.
+    """
+    soup = bs4.BeautifulSoup(html_string.encode('utf-8'), 'html.parser')
+
+    # Remove images with empty filepath.
+    for image in soup.findAll(name='oppia-noninteractive-image'):
+        if (not image.has_attr('filepath-with-value') or
+                image['filepath-with-value'] == '' or
+                image['filepath-with-value'] == '&quot;&quot;'):
+            image.decompose()
+            continue
+
+    # To add dimensions to images inside the collapsible component.
+    for collapsible_component in soup.findAll(
+            name='oppia-noninteractive-collapsible'):
+        if (not collapsible_component.has_attr('content-with-value') or
+                collapsible_component['content-with-value'] == ''):
+            continue
+
+        collapsible_component_html_string = unescape_html(
+            collapsible_component['content-with-value'])
+        collapsible_component_soup = bs4.BeautifulSoup(
+            collapsible_component_html_string, 'html.parser')
+        _modify_image_filename(
+            file_system, collapsible_component_soup)
+        collapsible_component['content-with-value'] = (
+            escape_html(python_utils.UNICODE(collapsible_component_soup)
```

Comment:
I'm a little concerned about all these random-looking replacements because I can't understand the logic for why they are correct (there is no explanation). Maybe add some comments describing why we're doing this.

Have we done things like this elsewhere before? Perhaps we could pull out one or more common standalone functions that take a string and (say ) delete quotes and replace br tags, and use that function in multiple places and give it a name that describes clearly what it is doing -- that will help with adding context, and also I see this sort of construction repeated in multiple places.